### PR TITLE
Cleanup DoNotUseWhenAllOrWaitAllWithSingleArgument

### PR DESCRIPTION
**Pull request by Youssef1313**
Quick follow-up to #4841

- Removed unnecessary derived classes.
- Used non-localized equivalence key.
- Renamed all contexts to `context.
- Used SymbolEqualityComparer.Default

cc: @ryzngard

### CODE REVIEW

**Comment by Youssef1313** [comment]
@mavasani I'm unable to debug the failure.

Locally, the tests fail with: `NuGet.Packaging.Core.PackagingException : The package is missing the required nuspec file. Path: C:\Users\Youssef\AppData\Local\Temp\test-packages\Microsoft.NETCore.App.Ref.3.1.0`

I have no clue how to fix this. :/

**Comment by Youssef1313** [comment]
> System.AggregateException : One or more errors occurred. (Unexpected Null)\n---- System.InvalidOperationException : Unexpected Null

:confused:

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Tasks/DoNotUseWhenAllOrWaitAllWithSingleArgument.cs [hunk 1]
```diff
@@ -42,28 +42,28 @@ public sealed override void Initialize(AnalysisContext context)
         {
             context.EnableConcurrentExecution();
             context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
-            context.RegisterCompilationStartAction(compilationContext =>
+            context.RegisterCompilationStartAction(context =>
```
Comment (ryzngard) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Tasks/DoNotUseWhenAllOrWaitAllWithSingleArgument.cs:45]: Is there any particular reason to rename this to just `context`?

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Tasks/DoNotUseWhenAllOrWaitAllWithSingleArgumentFixer.cs [hunk 2]
```diff
@@ -12,7 +13,8 @@
 
 namespace Microsoft.NetCore.Analyzers.Tasks
 {
-    public abstract class DoNotUseWhenAllOrWaitAllWithSingleArgumentFixer : CodeFixProvider
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
```
Comment (ryzngard) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Tasks/DoNotUseWhenAllOrWaitAllWithSingleArgumentFixer.cs:16]: oops 😅 I really thought I had this change in already. Thanks for fixing!

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Tasks/DoNotUseWhenAllOrWaitAllWithSingleArgument.cs [hunk 3]
```diff
@@ -112,7 +112,8 @@ private static bool IsSingleTaskArgument(IInvocationOperation invocation, INamed
                 return false;
             }
 
-            return namedTypeSymbol.Equals(taskType) || namedTypeSymbol.ConstructedFrom.Equals(genericTaskType);
+            return namedTypeSymbol.Equals(taskType, SymbolEqualityComparer.Default) ||
```
Comment (ryzngard) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Tasks/DoNotUseWhenAllOrWaitAllWithSingleArgument.cs:115]: I thought `symbol.Equals(otherSymbol)` implicitly used the `SymbolEqualityComparer.Default` if available?

**Comment by ryzngard** [comment]
> @mavasani I'm unable to debug the failure.
> 
> ~Locally, the tests fail with: `NuGet.Packaging.Core.PackagingException : The package is missing the required nuspec file. Path: C:\Users\Youssef\AppData\Local\Temp\test-packages\Microsoft.NETCore.App.Ref.3.1.0`~
> 
> ~I have no clue how to fix this. :/~ Deleting test-packages folder fixed it.
> 
> But the behavior of IOperation API looks very strange:
> 
> image
> 
> Is that a bug in the operation tree?

It looks like you figured this out, but the Value being an implicitly generated IArrayCreationOperation changes the behavior slightly.

**Comment by ryzngard** [comment]
Thanks for doing this!

**Review by ryzngard — APPROVED** [review]

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Tasks/DoNotUseWhenAllOrWaitAllWithSingleArgument.cs (hunk 3 above)
Comment (Youssef1313) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Tasks/DoNotUseWhenAllOrWaitAllWithSingleArgument.cs:115]: @ryzngard It does. But there is an analyzer (RS1024) to enforce using SymbolEqualityComparer explicitly to make the intent clear.

Currently the analyzer is disabled due to few bugs and will be re-enabled once they are fixed.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Tasks/DoNotUseWhenAllOrWaitAllWithSingleArgument.cs (hunk 1 above)
Comment (Youssef1313) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Tasks/DoNotUseWhenAllOrWaitAllWithSingleArgument.cs:45]: @ryzngard It's almost always incorrect to use a context other than the most inner one. So using the same name ensures this doesn't happen accidentally.

**Pull request merged**